### PR TITLE
Test e2e ginkgo

### PR DESCRIPTION
**Pull request by fran2810**
- Implemented e2e tests using the Ginkgo framework.
- Introduced utility and compose functions to streamline future test development.

### CODE REVIEW

File: test/k8s_tests/e2e_test.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+		newCapp := utilst.CreateCapp(K8sClient, baseCapp)
+		By("Checks unique creation of Capp")
+		Expect(newCapp.Name).ShouldNot(Equal(baseCapp.Name))
+		newCapp.Spec.ScaleMetric = "rps"
+		utilst.UpdateCapp(K8sClient, newCapp)
+		By("Checks if Capp Updated successfully")
+		Eventually(func() string {
+			newCapp = utilst.GetCapp(K8sClient, newCapp.Name, newCapp.Namespace)
+			return newCapp.Spec.ScaleMetric
+		}, 16*time.Second, 2*time.Second).Should(Equal("rps"), "Should fetch capp")
```
Comment (mzeevi) [inline test/k8s_tests/e2e_test.go:45]: Maybe move the `16*time.Second`, `2*time.Second`  to be consts that can be re-used?

File: test/k8s_tests/e2e_test.go [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+		utilst.DeleteCapp(K8sClient, newCapp)
+		By("Checks if deleted successfully")
+		Expect(utilst.DoesResourceExists(K8sClient, newCapp)).Should(BeFalse())
+	})
+})
```
Comment (mzeevi) [inline test/k8s_tests/e2e_test.go:50]: Add a blank space between `By` stanzas for better readability.

File: test/k8s_tests/mocks/base.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+						PodSpec: corev1.PodSpec{
+							Containers: []corev1.Container{
+								{
+									Env: []corev1.EnvVar{
+										{
+											Name:  "APP_NAME",
+											Value: "capp-default-test",
+										},
+									},
+									Image:     "sahar2339/example-python-app:v1-flask",
```
Comment (mzeevi) [inline test/k8s_tests/mocks/base.go:40]: Change image to use `quay.io/danateamorg/example-python-app:v1-flask` or alternatively upload it to `ghcr.io` in this repository and use that instead

File: test/k8s_tests/mocks/base.go [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+import (
+	rcsv1alpha1 "github.com/dana-team/container-app-operator/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	knativev1 "knative.dev/serving/pkg/apis/serving/v1"
+)
+
+var (
+	NsName = "capp-e2e-tests"
```
Comment (mzeevi) [inline test/k8s_tests/mocks/base.go:11]: Shouldn't it be `nsName`?

File: test/k8s_tests/test_suite.go [hunk 5]
```diff
@@ -15,17 +16,14 @@ import (
...
-	"testing"
 	"time"
-
-	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 var (
-	k8sClient client.Client
-	nsName    = "capp-e2e-tests"
+	K8sClient client.Client
```
Comment (mzeevi) [inline test/k8s_tests/test_suite.go:26]: is the capital letter here on purpose?

File: test/k8s_tests/utils/capp_adapter.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+const charset = "abcdefghijklmnopqrstuvwxyz0123456789"
+const RandStrLength = 10
+
+var seededRand *rand.Rand = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+func generateRandomString(length int) string {
```
Comment (mzeevi) [inline test/k8s_tests/utils/capp_adapter.go:18]: Add a short comment above the function

File: test/k8s_tests/utils/capp_adapter.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+// GetCapp fetch existing and return an instance of Capp.
+func GetCapp(k8sClient client.Client, name string, namespace string) *rcsv1alpha1.Capp {
+	capp := &rcsv1alpha1.Capp{
+		Spec: rcsv1alpha1.CappSpec{
+			RouteSpec: rcsv1alpha1.RouteSpec{},
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "test-capp",
+			Namespace: "test-ns",
```
Comment (mzeevi) [inline test/k8s_tests/utils/capp_adapter.go:56]: Are these values hard-coded on purpose?

File: test/k8s_tests/utils/resources_adpater.go [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import (
+	"context"
+	"fmt"
+	. "github.com/onsi/ginkgo/v2"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// DoesResourceExists checks if a given Kubernetes object exists in the cluster.
+func DoesResourceExists(k8sClient client.Client, obj client.Object) bool {
```
Comment (mzeevi) [inline test/k8s_tests/utils/resources_adpater.go:12]: `DoesResourceExist`, not `...Exists`

**Review by mzeevi — CHANGES_REQUESTED** [review]
Looks good overall but I have some questions

File: test/k8s_tests/mocks/base.go (hunk 4 above)
Comment (fran2810) [reply test/k8s_tests/mocks/base.go:11]: It won't be exposed

File: test/k8s_tests/test_suite.go (hunk 5 above)
Comment (fran2810) [reply test/k8s_tests/test_suite.go:26]: No I'll change it

File: test/k8s_tests/utils/capp_adapter.go (hunk 7 above)
Comment (fran2810) [reply test/k8s_tests/utils/capp_adapter.go:56]: No I'll change it

**Comment by mzeevi** [comment]
The CI still fails, think it's more related to the CI and not the tests

**Review by mzeevi — APPROVED** [review]

**Pull request merged**